clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -551,7 +572,8 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
                 }
 
                 /* if channel is ready write to any sockets that have space in their buffer and for which we have data */
-                if (channel.ready() && key.isWritable()) {
+                if (channel.ready() && key.isWritable() && !channel.maybeBeginClientReauthentication(
+                    () -> channelStartTimeNanos != 0 ? channelStartTimeNanos : time.nanoseconds())) {
```

Comment:
There was a desire to not call `time.nanoseconds()` unnecessarily.  We had a time value already, but it wasn't guaranteed to be current, so we have to get it again; making it a `Supplier` meands we can delay getting it for as long as possible and avoid it completely if the logic is short-circuited first.

Reply:
Does it really need to be that uptodate? Couldn't we just use the time that we have?

Reply:
Yeah, we can.  I mistakenly thought `currentTimeNanos` represented the time when `poll()` was called, but it is actually the moment when the `select()` call returns, so it is relatively recent.  So now I use the most recent value we have if there is one, otherwise we use `currentTimeNanos`.  I kept it as a `Supplier<>` though -- for symmetry with `KafkaChannel.maybeBeginServerReauthentication()`.  I can pass the long value in directly if you prefer, otherwise I think we're good here now.